File: templates/go/src/handlers/items.go
package handlers

import (
	"database/sql"
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"go-api/src/database"
	"go-api/src/models"
)

// ListItems returns all items
func ListItems(c *gin.Context) {
	rows, err := database.DB.Query("SELECT id, name, description, created_at FROM items ORDER BY id")
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}
	defer rows.Close()

	var items []models.Item
	for rows.Next() {
		var item models.Item
		err := rows.Scan(&item.ID, &item.Name, &item.Description, &item.CreatedAt)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
		items = append(items, item)
	}

	c.JSON(http.StatusOK, items)
}

// GetItem returns a single item by ID
func GetItem(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid ID"})
		return
	}

	var item models.Item
	err = database.DB.QueryRow(
		"SELECT id, name, description, created_at FROM items WHERE id = $1",
		id,
	).Scan(&item.ID, &item.Name, &item.Description, &item.CreatedAt)

	if err == sql.ErrNoRows {
		c.JSON(http.StatusNotFound, gin.H{"error": "Item not found"})
		return
	}
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusOK, item)
}

// CreateItem creates a new item
func CreateItem(c *gin.Context) {
	var req models.CreateItemRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	var item models.Item
	err := database.DB.QueryRow(
		"INSERT INTO items (name, description) VALUES ($1, $2) RETURNING id, name, description, created_at",
		req.Name, req.Description,
	).Scan(&item.ID, &item.Name, &item.Description, &item.CreatedAt)

	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	c.JSON(http.StatusCreated, item)
}

// DeleteItem deletes an item by ID
func DeleteItem(c *gin.Context) {
	id, err := strconv.Atoi(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid ID"})
		return
	}

	result, err := database.DB.Exec("DELETE FROM items WHERE id = $1", id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
		return
	}

	if rowsAffected == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "Item not found"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"message": "Item deleted successfully",
		"id":      id,
	})
}

File: templates/rust/src/handlers/items.rs
use actix_web::{web, HttpResponse, Responder, Result};
use sqlx::PgPool;
use chrono::DateTime;
use chrono::Utc;

use crate::models::{Item, CreateItem};

pub async fn list_items(pool: web::Data<PgPool>) -> Result<impl Responder> {
    let rows = sqlx::query("SELECT id, name, description, created_at FROM items ORDER BY id")
        .fetch_all(pool.get_ref())
        .await
        .map_err(|e| {
            eprintln!("Database error: {}", e);
            actix_web::error::ErrorInternalServerError("Database error")
        })?;

    let items: Vec<Item> = rows
        .iter()
        .map(|row| Item {
            id: Some(row.get("id")),
            name: row.get("name"),
            description: row.get("description"),
            created_at: row.get::<Option<DateTime<Utc>>, _>("created_at")
                .map(|dt| dt.format("%Y-%m-%d %H:%M:%S").to_string()),
        })
        .collect();

    Ok(HttpResponse::Ok().json(items))
}

pub async fn get_item(pool: web::Data<PgPool>, path: web::Path<i32>) -> Result<impl Responder> {
    let id = path.into_inner();
    
    let row = sqlx::query("SELECT id, name, description, created_at FROM items WHERE id = $1")
        .bind(id)
        .fetch_optional(pool.get_ref())
        .await
        .map_err(|e| {
            eprintln!("Database error: {}", e);
            actix_web::error::ErrorInternalServerError("Database error")
        })?;

    match row {
        Some(row) => {
            let item = Item {
                id: Some(row.get("id")),
                name: row.get("name"),
                description: row.get("description"),
                created_at: row.get::<Option<DateTime<Utc>>, _>("created_at")
                    .map(|dt| dt.format("%Y-%m-%d %H:%M:%S").to_string()),
            };
            Ok(HttpResponse::Ok().json(item))
        }
        None => Ok(HttpResponse::NotFound().json(serde_json::json!({
            "error": "Item not found"
        }))),
    }
}

pub async fn create_item(
    pool: web::Data<PgPool>,
    item: web::Json<CreateItem>,
) -> Result<impl Responder> {
    let result = sqlx::query(
        "INSERT INTO items (name, description) VALUES ($1, $2) RETURNING id, name, description, created_at"
    )
    .bind(&item.name)
    .bind(&item.description)
    .fetch_one(pool.get_ref())
    .await
    .map_err(|e| {
        eprintln!("Database error: {}", e);
        actix_web::error::ErrorInternalServerError("Database error")
    })?;

    let created_item = Item {
        id: Some(result.get("id")),
        name: result.get("name"),
        description: result.get("description"),
        created_at: result.get::<Option<DateTime<Utc>>, _>("created_at")
            .map(|dt| dt.format("%Y-%m-%d %H:%M:%S").to_string()),
    };

    Ok(HttpResponse::Created().json(created_item))
}

pub async fn delete_item(pool: web::Data<PgPool>, path: web::Path<i32>) -> Result<impl Responder> {
    let id = path.into_inner();
    
    let result = sqlx::query("DELETE FROM items WHERE id = $1")
        .bind(id)
        .execute(pool.get_ref())
        .await
        .map_err(|e| {
            eprintln!("Database error: {}", e);
            actix_web::error::ErrorInternalServerError("Database error")
        })?;

    if result.rows_affected() > 0 {
        Ok(HttpResponse::Ok().json(serde_json::json!({
            "message": "Item deleted successfully",
            "id": id
        })))
    } else {
        Ok(HttpResponse::NotFound().json(serde_json::json!({
            "error": "Item not found"
        })))
    }
}

